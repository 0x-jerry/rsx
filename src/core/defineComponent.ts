import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
