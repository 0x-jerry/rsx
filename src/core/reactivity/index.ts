export { computed, reactive, ref } from '@vue/reactivity'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './utils'
export * from './watch'
