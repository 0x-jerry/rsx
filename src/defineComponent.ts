import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
