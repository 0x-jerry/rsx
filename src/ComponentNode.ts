import { camelCase, isObject, PascalCase } from '@0x-jerry/utils'
import {
  appendToCurrentContext,
  createNodeContext,
  type DNodeContext,
  popCurrentContext,
  setCurrentContext,
} from './context'
import type { FunctionalComponent } from './defineComponent'
import { isHTMLNode } from './node'
import type { AnyProps } from './props'
import { isRef, unref } from './reactivity'

let componentId = 0

export class ComponentNode {
  readonly __cf = true

  id = componentId++
  type: FunctionalComponent
  props?: AnyProps
  children: unknown[]

  instance!: DNodeContext

  _initialized = false

  constructor(
    type: FunctionalComponent,
    props: AnyProps | undefined,
    children: unknown[],
  ) {
    this.type = type
    this.props = props
    this.children = children
  }

  initialize() {
    if (this._initialized) {
      console.error('[ComponentNode] ComponentNode has been initialized')
      return
    }

    this.instance = this._createComponentInstance()
  }

  _createComponentInstance() {
    const ctx = createNodeContext(this.type.name)

    appendToCurrentContext(ctx)

    setCurrentContext(ctx)

    const rootEl = this.type(this.props, this.children)

    if (isComponentNode(rootEl)) {
      rootEl.initialize()

      ctx.el = rootEl.instance.el
    } else if (isHTMLNode(rootEl)) {
      ctx.el = rootEl as ChildNode
    } else {
      console.warn('[ComponentNode] Invalid component node', rootEl)
    }

    popCurrentContext()

    return ctx
  }
}

export function createComponentNode(
  type: FunctionalComponent,
  props: AnyProps | undefined,
  children: unknown[],
) {
  const node = new ComponentNode(type, props, children)

  return node
}

export function isComponentNode(o: unknown): o is ComponentNode {
  return isObject(o) && '__cf' in o && o.__cf === true
}

export function transformProps(type: any, props?: AnyProps): any {
  const _raw: Record<string, any> = {}

  if (!props) return _raw

  Object.entries(props).forEach(([key, value]) => {
    if (!key.startsWith('$')) {
      // Prevent change props directly.
      _raw[key] = value
      return
    }

    if (key === '$') {
      const newProps = transformDefaultBinding(type, value, props)

      // todo, check duplicate key
      Object.assign(_raw, newProps)
      return
    }

    // normal binding syntax sugar
    // <input $value:trim={refValue} />

    const [name, modifier] = key.slice(1).split(':')
    _raw[camelCase(name)] = value

    if (isRef(value)) {
      // fix me: compose events
      _raw[`onUpdate${PascalCase(name)}`] = (v: unknown) => {
        value.value = v
      }
    }
  })

  // return _raw
  return new Proxy(_raw, {
    get(t, p, r) {
      return unref(Reflect.get(t, p, r))
    },
    set() {
      // Prevent change props directly.
      return true
    },
  })
}

// todo, default binding syntax sugar
// current: $xx={refValue}
// support new sugar: $xx={[data, 'key']}
function transformDefaultBinding(
  type: any,
  value: any,
  allProps: Record<string, any>,
) {
  const props: Record<string, any> = {}

  if (type === 'input') {
    if (unref(allProps.type) === 'checkbox') {
      props.checked = value
      if (isRef(value)) {
        props.onChange = (e: InputEvent) => {
          value.value = (e.target as HTMLInputElement).checked
        }
      }
    } else {
      props.value = value

      if (isRef(value)) {
        props.onInput = (e: InputEvent) => {
          value.value = (e.target as HTMLInputElement).value
        }
      }
    }
  } else if (type === 'select') {
    props.value = value

    if (isRef(value)) {
      props.onChange = (e: InputEvent) => {
        value.value = (e.target as HTMLSelectElement).value
      }
    }
  }

  return props
}
