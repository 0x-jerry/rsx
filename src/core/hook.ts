import type { Fn } from '@0x-jerry/utils'
import type { WatchCallback, WatchEffect, WatchSource } from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  getCurrentContext,
} from './context'
import { type WatchHandle, type WatchOptions, watch } from './reactivity'

export function unmount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeUnmount)

  ctx.el?.remove()

  ctx.children?.forEach((child) => unmount(child))

  ctx.emit(DNodeContextEventName.unmounted)
}

export function mount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeMount)

  ctx.children?.forEach((item) => mount(item))

  ctx.emit(DNodeContextEventName.mounted)
}

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

export function onBeforeMount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeMount, fn)
}

export function onMounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.mounted, fn)
}

export function onBeforeUnmount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeUnmount, fn)
}

export function onUnmounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.unmounted, fn)
}

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stop = watch(getter, fn, opt)

  onUnmounted(stop)

  return stop
}
