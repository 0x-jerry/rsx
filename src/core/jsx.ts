import { camelCase, isString, PascalCase } from '@0x-jerry/utils'
import { isRef, unref } from '@vue/reactivity'
import {
  appendToCurrentContext,
  createNodeContext,
  popCurrentContext,
  setCurrentContext,
} from './context'
import {
  createFragment,
  createNativeElement,
  type DComponent,
  type DNode,
  isDComponent,
} from './node'

type FunctionalComponent = (props?: any, children?: DNode[]) => DComponent

export function h(
  type: string | FunctionalComponent | DComponent,
  props?: Record<string, any>,
  ...children: DNode[]
): DComponent {
  if (isDComponent(type)) {
    return type
  }

  const _props = transformProps(type, props)

  if (type === Fragment) {
    return type(_props, children)
  }

  if (isString(type)) {
    return createNativeElement(type, _props, children)
  }

  return createComponentInstance(type, _props, children)
}

export const Fragment: FunctionalComponent = (_, children) => {
  return createFragment(children || [])
}

function createComponentInstance(
  type: FunctionalComponent,
  props?: any,
  children?: DNode[],
) {
  const ctx = createNodeContext(type.name)

  appendToCurrentContext(ctx)

  setCurrentContext(ctx)

  const el = type(props, children)

  el._ = ctx

  popCurrentContext()
  return el
}

function transformProps(type: any, props?: Record<string, any>): any {
  const _raw: Record<string, any> = {}

  if (!props) return _raw

  Object.entries(props).forEach(([key, value]) => {
    if (!key.startsWith('$')) {
      // Prevent change props directly.
      _raw[key] = value
      return
    }

    if (key === '$') {
      const newProps = transformDefaultBinding(type, value, props)

      // todo, check duplicate key
      Object.assign(_raw, newProps)
      return
    }

    // normal binding syntax sugar
    // <input $value:trim={refValue} />

    const [name, modifier] = key.slice(1).split(':')
    _raw[camelCase(name)] = value

    if (isRef(value)) {
      // fix me: compose events
      _raw[`onUpdate${PascalCase(name)}`] = (v: unknown) => (value.value = v)
    }
  })

  // return _raw
  return new Proxy(_raw, {
    get(t, p, r) {
      return unref(Reflect.get(t, p, r))
    },
    set() {
      // Prevent change props directly.
      return true
    },
  })
}

// todo, default binding syntax sugar
// current: $xx={refValue}
// suppport new sugar: $xx={[data, 'key']}
function transformDefaultBinding(
  type: any,
  value: any,
  allProps: Record<string, any>,
) {
  const props: Record<string, any> = {}

  if (type === 'input') {
    if (unref(allProps.type) === 'checkbox') {
      props.checked = value
      if (isRef(value)) {
        props.onChange = (e: InputEvent) => {
          value.value = (e.target as HTMLInputElement).checked
          console.log('checked', value.value)
        }
      }
    } else {
      props.value = value

      if (isRef(value)) {
        props.onInput = (e: InputEvent) =>
          (value.value = (e.target as HTMLInputElement).value)
      }
    }
  } else if (type === 'select') {
    props.value = value

    if (isRef(value)) {
      props.onChange = (e: InputEvent) =>
        (value.value = (e.target as HTMLSelectElement).value)
    }
  }

  return props
}
