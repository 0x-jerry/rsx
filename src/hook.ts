import type { EmptyObject, Fn } from '@0x-jerry/utils'
import {
  readonly,
  type WatchCallback,
  type WatchEffect,
  type WatchSource,
} from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  type DNodeEventMap,
  getCurrentContext,
} from './context'
import { isRef, type WatchHandle, type WatchOptions, watch } from './reactivity'

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

function createContextHook(name: keyof DNodeEventMap) {
  return (fn: Fn, ctx?: DNodeContext) => {
    ctx ||= useContext()

    ctx.on(name, fn)
  }
}

export const onBeforeMount = createContextHook(
  DNodeContextEventName.beforeMount,
)
export const onMounted = createContextHook(DNodeContextEventName.mounted)

export const onBeforeUnmount = createContextHook(
  DNodeContextEventName.beforeUnmount,
)
export const onUnmounted = createContextHook(DNodeContextEventName.unmounted)

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stopHandle = watch(getter, fn, opt)

  onUnmounted(stopHandle)

  return stopHandle
}

export type InjectionKey<T> = (string | symbol) & { _: T }

export function provide<T = unknown>(
  key: string | symbol | InjectionKey<T>,
  value: T,
) {
  const ctx = useContext()
  ctx.ex ||= {}
  ctx.ex[key] = value
}

export function inject<T>(
  key: string | symbol | InjectionKey<T>,
): T | undefined {
  const ctx = useContext()

  let node = ctx.parent

  while (node) {
    if (node.ex?.[key]) {
      return node.ex[key] as T
    }

    node = node.parent
  }
}

export function useRawProps() {
  const ctx = useContext()

  return ctx._node?.props || {}
}

export function useExpose<T extends EmptyObject>(exposed: T) {
  const ref = useRawProps().ref

  if (isRef(ref)) {
    ref.value = readonly(exposed)
  }
}

export function useRawChildren() {
  return useContext()._node?.children || []
}
