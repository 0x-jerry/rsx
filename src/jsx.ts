import { isString } from '@0x-jerry/utils'
import type { FunctionalComponent } from './defineComponent'
import { Fragment } from './internalComponents'
import { createFragment } from './nodes/AnchorNode'
import { createComponentNode } from './nodes/ComponentNode'
import { createNativeNode } from './nodes/NativeNode'
import type { NodeElement } from './nodes/shared'
import type { AnyProps } from './props'

export function h(
  type: string | FunctionalComponent,
  props?: AnyProps,
  ...children: unknown[]
): NodeElement {
  if (isString(type)) {
    return createNativeNode(type, props, children)
  }

  if (type === Fragment) {
    return createFragment(children)
  }

  return createComponentNode(type, props, children)
}
