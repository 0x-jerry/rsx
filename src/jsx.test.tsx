/** biome-ignore-all lint/style/noNonNullAssertion: test case */

import { isComponentNode } from './ComponentNode'
import { dc } from './defineComponent'
import { useWatch } from './hook'
import { h } from './jsx'
import { $, nextTick, ref } from './reactivity'
import { contextToJson, defineComponentName, mountTestApp } from './test'

describe('jsx', () => {
  it('dom element', () => {
    const el = h('div')

    expect(el).instanceof(HTMLDivElement)
  })

  it('component node', () => {
    const Comp = () => h('div', { class: 'test' })

    const node = h(Comp)

    expect(isComponentNode(node)).toBe(true)
  })

  it('should set props when create', () => {
    const Comp = () => h('div', { class: 'test' })

    const el = mountTestApp(Comp)

    expect(el.querySelector('.test')).instanceOf(HTMLDivElement)
  })

  it('should update ref props', async () => {
    const v = ref(0)

    const Comp = () => h('div', { 'data-x': v, class: 'test' })

    const el = mountTestApp(Comp)

    expect(el.querySelector('.test')?.getAttribute('data-x')).toBe('0')
    v.value++

    await nextTick()
    expect(el.querySelector('.test')?.getAttribute('data-x')).toBe('1')
  })

  it('should set event listener when create', () => {
    const fn = vi.fn()

    const Comp = () => h('div', { onClick: fn, class: 'test' })

    const el = mountTestApp(Comp).querySelector('.test')

    el?.dispatchEvent(new Event('click'))
    el?.dispatchEvent(new Event('click'))
    expect(fn).toBeCalledTimes(2)
  })

  describe('handle $ binding prop', () => {
    it('should handle $ prop on input[text] element', () => {
      const v = ref('123')

      const Comp = () => h('input', { $: v })

      const el = mountTestApp(Comp).querySelector('input')!

      expect(el.value).toBe('123')

      el.value = '12344'
      expect(v.value).toBe('123')

      el.dispatchEvent(new Event('input'))
      expect(v.value).toBe('12344')
    })

    it('should handle $ prop on input[checkbox] element', () => {
      const v = ref(true)

      const Comp = () => h('input', { type: 'checkbox', $: v })

      const el = mountTestApp(Comp).querySelector('input')!

      expect(el.checked).toBe(true)

      el.checked = false
      expect(v.value).toBe(true)

      el.dispatchEvent(new Event('change'))
      expect(v.value).toBe(false)
    })

    it('should handle $ prop on select element', () => {
      const v = ref('1')

      const Comp = () => (
        <select $={v}>
          <option value="1"></option>
          <option value="2"></option>
        </select>
      )

      const el = mountTestApp(Comp).querySelector('select')!

      el.value = '2'
      expect(v.value).toBe('1')

      el.dispatchEvent(new Event('change'))
      expect(v.value).toBe('2')
    })
  })

  it('should work without context when render with non-reactive props', () => {
    const el = h('div', { class: 'test' }) as HTMLDivElement

    expect(el.classList.contains('test')).toBe(true)
  })

  it('should only work with context when render with reactive props', () => {
    const fn = () => {
      const cls = ref('test')

      h('div', { class: cls })
    }

    expect(fn).throw('This should only used inside functional component')
  })

  it('should render as text node', async () => {
    const v = ref('hello')

    const Comp = () => h('div', {}, v)

    const el = mountTestApp(Comp)

    const text = el.firstChild!.childNodes.item(0)
    expect(text).instanceOf(Text)
    expect(text.textContent).toBe('hello')

    v.value = 'world'
    expect(text.textContent).toBe('hello')
    await nextTick()
    expect(text.textContent).toBe('world')
  })

  it('should not allowed to change props directly', async () => {
    type Props = {
      $v: string
    }

    const Comp = dc<Props>((props) => {
      // should not work
      props.v = '123'

      return h(
        'div',
        {
          onClick() {
            props.onUpdateV?.('123')
          },
        },
        $(props, 'v'),
      )
    })

    const props = {
      v: ref('1'),
      onUpdateV(v: string) {
        props.v.value = v
      },
    }

    const el = mountTestApp(() => h(Comp, props)).firstChild!

    expect(el.textContent).toBe('1')
    expect(props.v.value).toBe('1')

    el.dispatchEvent(new Event('click'))

    expect(props.v.value).toBe('123')
    expect(el.textContent).toBe('1')
    await nextTick()
    expect(el.textContent).toBe('123')
  })

  it('should remain reactive when passing a ref value', async () => {
    type Props = {
      v: number
    }

    const fn = vi.fn()

    const Comp = dc<Props>((props) => {
      useWatch(() => props.v, fn)

      return h('div')
    })

    const props = {
      v: ref(1),
      x: 1,
    }

    mountTestApp(() => h(Comp, props))

    await nextTick()
    expect(fn).toBeCalledTimes(0)

    props.v.value++
    await nextTick()
    expect(fn).toBeCalledTimes(1)
  })
})

describe('jsx context tree', () => {
  it('context tree', () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const B = dc((_, children) => (
      <div class="B">
        <span>b</span>
        {children}
      </div>
    ))

    defineComponentName(B, 'B')

    const App = dc(() => (
      <A>
        <span>1</span>
        <B>
          <span>2</span>
          <A></A>
        </B>
        <A></A>
        <span>3</span>
      </A>
    ))

    defineComponentName(App, 'App')

    const root = mountTestApp(App)

    expect(root).toMatchSnapshot('html')

    const ctxTree = contextToJson(root._)

    expect(ctxTree).toMatchSnapshot('ctx tree')
  })
})
