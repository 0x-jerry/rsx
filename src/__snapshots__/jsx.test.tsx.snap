// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx context tree > context tree > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "name": "A",
            },
          ],
          "name": "B",
        },
        {
          "children": [],
          "name": "A",
        },
      ],
      "name": "A",
    },
  ],
  "name": "Root",
}
`;

exports[`jsx context tree > context tree > html 1`] = `
<div
  id="app"
>
  <div
    class="A"
  >
    <span>
      a
    </span>
    <span>
      1
    </span>
    <div
      class="B"
    >
      <span>
        b
      </span>
      <span>
        2
      </span>
      <div
        class="A"
      >
        <span>
          a
        </span>
      </div>
    </div>
    <div
      class="A"
    >
      <span>
        a
      </span>
    </div>
    <span>
      3
    </span>
  </div>
</div>
`;
